src/Features/CSharp/Portable/CodeRefactorings/UseRecursivePatterns/UseRecursivePatternsCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,438 @@
+                !TryGetParentNames(rightReceiver, out var rightNames))
+            {
+                return null;
+            }
+
+            return (leftReceiver, leftNames, rightNames);
+        }
+
+        /// <summary>
+        /// Walks up the receiver and collect any names on the right-hand-side. 
+        /// These are already verified in <see cref="GetInnermostReceiver(ExpressionSyntax, SemanticModel)"/> below to have a proper name.
+        /// </summary>
+        private static bool TryGetParentNames(ExpressionSyntax node, out ImmutableArray<IdentifierNameSyntax> names)
+        {
+            using var _ = ArrayBuilder<IdentifierNameSyntax>.GetInstance(out var builder);
+            CollectNames(node.Parent, builder);
+            names = builder.ToImmutableOrNull();
+            return !names.IsDefault;
+
+            // We collect names inside out, so that the outermost name ends up in the innermost subpattern.
+            static void CollectNames(SyntaxNode? node, ArrayBuilder<IdentifierNameSyntax> builder)
+            {
+                switch (node)
+                {
+                    case MemberAccessExpressionSyntax(SimpleMemberAccessExpression) { Name: IdentifierNameSyntax name } memberAccess:
+                        CollectNames(memberAccess.Parent, builder);
+                        builder.Add(name);
+                        break;
+                    case ConditionalAccessExpressionSyntax { WhenNotNull: IdentifierNameSyntax name } memberAccess:
+                        CollectNames(memberAccess.Parent, builder);
+                        builder.Add(name);
+                        break;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Obtain the innermost receiver that is on the left of a member access which cannot be converted to a subpattern.
+        /// </summary>
+        private static ExpressionSyntax GetInnermostReceiver(ExpressionSyntax node, SemanticModel model)
+        {
+            return node switch
+            {
+                MemberAccessExpressionSyntax(SimpleMemberAccessExpression) { Name: IdentifierNameSyntax name } memberAccess
+                    when RequiresInstanceReceiver(name, model) => GetInnermostReceiver(memberAccess.Expression, model),
+                ConditionalAccessExpressionSyntax { WhenNotNull: IdentifierNameSyntax name } memberAccess
+                    when RequiresInstanceReceiver(name, model) => GetInnermostReceiver(memberAccess.Expression, model),
+                _ => node,
+            };
+
+            static bool RequiresInstanceReceiver(SyntaxNode node, SemanticModel model)
+                => model.GetSymbolInfo(node).Symbol?.IsStatic == false;
+        }
+
+        private sealed class MyCodeAction : CodeActions.CodeAction.DocumentChangeAction
+        {
+            public MyCodeAction(string title, Func<CancellationToken, Task<Document>> createChangedDocument)
+                : base(title, createChangedDocument)
```

Comment:
Needs equivalence key with value `nameof(TheTitleResourceAfterMovingTheTitleToResx)`

Reply:
I don't think that's necessary for a refactoring.

Reply:
In a recent PR, @sharwell has suggested doing this for refactorings. I don't remember if this PR had some special reason to do this or it was a general feedback. It's good to do anyway. But @sharwell can confirm and provide more context.